### ISSUE TITLE
Struct method which prints fields of the struct it is in fails to compile for some weird reason

### ISSUE THREAD

**Issue by yasayer** [ziglang/zig#7140]
The title sucks, but honestly, I haven't got any better idea. Anyways, consider the following code.

```zig
const std = @import("std");
const print = std.debug.print;

const Card = struct {
    const Self = @This();

    rank: i5,
    suit: i3,

    pub fn initRankAndSuit(rank: i5, suit: i3) Self {
        return Self{
            .rank = rank,
            .suit = suit,
        };
    }
};

const Deck = struct {
    const Self = @This();

    cards: [52]Card,

    pub fn init() Self {
        var self: Self = undefined;

        for (self.cards) |*card, i| {
            card.* = Card.initRankAndSuit(@intCast(i5, @rem(i, 13)), @intCast(i3, @divTrunc(i, 13)));
        }

        return self;
    }

    pub fn print(self: *Self) void {
        for (self.cards) |card, i| {
            if (i < 52) {
                print("{}\n", .{card});
            } else {
                print("{}\n\n", .{card});
            }
        }
    }
};

pub fn deckPrint(self: *Deck) void {
    for (self.cards) |card, i| {
        if (i < 52) {
            print("{}\n", .{card});
        } else {
            print("{}\n\n", .{card});
        }
    }
}
```

I suppose printing functions are the most important thing here, I've left the rest of this code so that you can get the context.

Now, I'm going to use the printing functions, starting off with the very first one.

```zig
pub fn main() !void {
    var deck = Deck.init();

    deckPrint(&deck); // works just fine!
}
```

Everything's fine, moving the next two.

```zig
pub fn main() !void {
    var deck = Deck.init();

    Deck.print(&deck); // note: cast discards const qualifier; error: expected type '*Deck', found '*const [3:0]u8'
    deck.print(); // note: cast discards const qualifier; error: expected type '*Deck', found '*const [3:0]u8'
}
```

Something went wrong here, and I'm not quite sure what it might be. Surprisingly...

```zig
pub fn main() !void {
    var deck = Deck.init();

    print("whatever\n", .{});
    Deck.print(&deck); // fine!
    deck.print(); // also fine!
}
```

...this is alright. It seems that these two functions fail to compile unless you precede them with some other print, whatever it'd print.

**Comment by LemonBoy** [comment]
The `print` variable inside `Deck.print` is bound to `Deck.print` itself and not to the imported function.
Calling `print` beforehand masks the problem as there's no ambiguity anymore.

Looks like a compiler bug.

**Comment by mikdusan** [comment]
This looks fixed to me as of 0.9.0 and master. Originally an ambiguity issue, and at 0.9.0 the compiler began to omit this error message which effectively forces the source code to be fixed:

```sh
$ ./zig-linux-x86_64-0.9.0/zig build-obj repro.zig
repro.zig:36:17: error: ambiguous reference
                print("{}\n", .{card});
                ^
repro.zig:33:9: note: declared here
    pub fn print(self: *Self) void {
        ^
repro.zig:2:1: note: also declared here
const print = std.debug.print;
^
```

The source code can be fixed by making `print` unambiguous; here is one way:
```zig
    pub fn print(self: *Self) void {
        for (self.cards) |card, i| {
            if (i < 52) {
                std.debug.print("{}\n", .{card});
            } else {
                std.debug.print("{}\n\n", .{card});
            }
        }
    }
```
